Mlem/API/APIRequest.swift

```diff
@@ -0,0 +1,31 @@
+//
+//  APIRequest.swift
+//  Mlem
+//
+//  Created by Nicholas Lawson on 06/06/2023.
+//
+
+import Foundation
+
+enum APIRequestError: Error {
+    case encoding
+}
+
+protocol APIRequest {
+    associatedtype Response: Decodable
+    
+    var endpoint: URL { get }
```

Comment:
Wouldn't it be better to make these immutable? Not sure, but since we're basically never editing an APIRequest after creation, swift could use the hint to maybe optimise the struct/handling?

Reply:
This is just a weird one with how protocols are defined in Swift, you cannot declare a `let` inside a protocol definition, you instead declare if the protocol requires to read `{ get }`, or read and write `{ get set }`.

So in this case we need to use the keyword `var`, but declare conformance only requires read `{ get }`, inside the actual concrete structs they are still represented as `let` this way

Reply:
Ah, in that case this should be fine

I dont know much about swift protocols, thus the suggestion